ext/crypto/00_crypto.js

```diff
@@ -898,7 +1112,274 @@
+              //2.
+              const keyInfo = parseRsaPkcs(keyData, "spki");
+              //3.
+              if (!keyInfo) {
+                throw new DOMException("Invalid spki data", "DataError");
+              }
+              //4.
+              let hash;
+              //5.
+              //const alg = keyInfo.algorithm;
+              //hash = match { OID }
+              //6.
+              // TODO(@SeanWykes): Check alg==rsaEncryptionOID
+              //7.
+              if (hash) {
+                const normalizedHash = normalizeAlgorithm(hash, "digest");
+
+                if (normalizedHash.name !== normalizedAlgorithm.hash) {
+                  throw new DOMException(
+                    "Mismatched Hash algorithm and spki algorithmIdentifier",
+                    "DataError",
+                  );
+                }
+              }
+              //8.
+              //const publicKey = keyData;
+              //9.
+              modulusLength = keyInfo.n.length * 8;
+              //10.
+              type = "public";
+              data = keyData;
+              break;
+            }
+            case "pkcs8": {
+              // 1.
+              if (
+                ArrayPrototypeFind(
+                  keyUsages,
+                  (u) => !ArrayPrototypeIncludes(["sign"], u),
+                ) !== undefined
+              ) {
+                throw new DOMException("Invalid key usages", "SyntaxError");
+              }
+              //2.
+              const keyInfo = parseRsaPkcs(keyData, "pkcs8");
+              //3.
+              if (!keyInfo) {
+                throw new DOMException("Invalid pkcs8 data", "DataError");
+              }
+              //4.
+              let hash;
+              //5.
+              //const alg = keyInfo.algorithm;
+              //hash = match { OID }
+              //6.
+              // TODO(@SeanWykes): Check alg==rsaEncryptionOID
+              //7.
+              if (hash) {
+                const normalizedHash = normalizeAlgorithm(hash, "digest");
+
+                if (normalizedHash.name !== normalizedAlgorithm.hash) {
+                  throw new DOMException(
+                    "Mismatched Hash algorithm and PKCS8 algorithmIdentifier",
+                    "DataError",
+                  );
+                }
+              }
+              //8.
+              //const privateKey = keyData;
+              //9.
+              modulusLength = keyInfo.n.length * 8;
+              //10.
+              type = "private";
+              data = keyData;
+              break;
+            }
+            default:
+              throw new DOMException("Not implemented", "NotSupportedError");
+          }
+
+          if (keyUsages.length == 0) {
+            throw new DOMException("Key usage is empty", "SyntaxError");
+          }
+
+          const handle = {};
+          WeakMapPrototypeSet(KEY_STORE, handle, {
+            type: (type == "private") ? "pkcs8" : "spki",
```

Comment:
This won't work. The key for a particular algorithm must be same throughout all the operations.

Reply:
Sorry, I can't see the problem here. The code imports jwk, converting private to 'pkcs8' format and public to spki format for  use of the other op_crypto operations. Please elucidate.

Reply:
Other RSA operations assume that the key is of type raw i.e `pkcs1` format. 
An `spki`/`pkcs8` key will fail eg: https://github.com/denoland/deno/blob/a95ca9dc70515240b1be7a12dbf686b1ebfb3490/ext/crypto/lib.rs#L275

Reply:
ok, PR#11880 altered this .. I had patched the operations to match both spki/pkcs8 .. will rework it.

Reply:
Instead of having mixed type keys in storage. Can you just call `.to_pkcs1_der()` on the private key (once you have implemented the Rust op) and store it as pkcs1? It will make life much easier for the maintenance and eventually migrating to a new crypto storage.

Reply:
reworked and merged. 

There will still be a problem with importing RSA Public keys via SPKI/JWK, since I am now using the KeyFormat enum to differentiate between private (Raw) and public (Spki). Ideally we would store public keys in Pkcs1 format, I think, so maybe need to include a new property in KeyData that indicates private/public. What do you think?

Reply:
The `CryptoKey` object already has the `type` property which is associated to the key handle.

Reply:
Sure, but that currently doesn't reach rust ops, for example, op_crypto_verify_key receives a VerifyArg with key: KeyData object but no CryptoKey. Verify must be able to distinguish between imported private and public keys.

Reply:
Ah yes good point, we'll need to pass the `type` to these particular rust ops. It probably wasn't done before we weren't dealing with public keys directly.

Reply:
Need to include a "keyType" property on KeyData, and copy from CryptoKey.type during import/generate .. ECDH and ECDSA will need this as well.